### PR TITLE
Attach 'send' autograd function to the autograd graph as part of RPC.

### PR DESCRIPTION
**Pull request by pritamdamania87**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#24876 Attach 'send' autograd function to the autograd graph as part of RPC.**

This contains very basic functionality of adding 'send' autograd
function to our autograd graph. The purpose of this change is to validate the
basic structure proposed here makes sense. Once this makes sense, we can build
upon this to address more complicated scenarios. At a high level we've added
the following functionality:

1) Define a very simple 'SendRpcBackwards' autograd function.
2) Attach this function to appropriate tensors when we call an RPC.
3) Store the send function in our distributed autograd context.

GitHub Issue: https://github.com/pytorch/pytorch/issues/23110
Differential Revision: [D16903255](https://our.internmc.facebook.com/intern/diff/D16903255/)

### CODE REVIEW

File: tools/build_variables.py
Comment (mrshenli) [inline tools/build_variables.py]: we are using snake names now

File: torch/csrc/distributed/rpc/python_functions.cpp
Comment (mrshenli) [inline torch/csrc/distributed/rpc/python_functions.cpp]: How does this work for `RRef` args?

File: torch/csrc/distributed/autograd/functions/SendRpcBackwards.h
Comment (mrshenli) [inline torch/csrc/distributed/autograd/functions/SendRpcBackwards.h]: Should this be called `SendRpcBackward` instead? Not 100% sure, just want to make sure we are consistent with other backward node.

File: torch/csrc/distributed/autograd/functions/SendRpcBackwards.h
Comment (mrshenli) [inline torch/csrc/distributed/autograd/functions/SendRpcBackwards.h]: We would also need a `RecvRpcBackward` to pair with this `SendRpcBackward`, no?

File: test/test_rpc.py
Comment (mrshenli) [inline test/test_rpc.py]: Is this just for temporary test purpose, or do we plan to keep this API in the final version?

File: torch/csrc/distributed/autograd/context/DistAutogradContext.cpp
Comment (mrshenli) [inline torch/csrc/distributed/autograd/context/DistAutogradContext.cpp]: Should this be a map, as recv need to use its ID to find the corresponding send??

**Comment by ezyang** [comment]
When you get a chance please rebase past master failures

**Comment by aazzolini** [comment]
@pritamdamania87 , how are we going to do this when the object containing the tensor is a python object?

File: torch/csrc/distributed/rpc/python_functions.cpp
Comment (pritamdamania87) [reply torch/csrc/distributed/rpc/python_functions.cpp]: We discussed this offline, as part of .to_here() we would attach the necessary args.

File: torch/csrc/distributed/autograd/functions/SendRpcBackwards.h
Comment (pritamdamania87) [reply torch/csrc/distributed/autograd/functions/SendRpcBackwards.h]: Yes, I wanted to keep this PR simple and added only `send` for early feedback on the general design. Will follow up on `recv` in a different PR.

File: torch/csrc/distributed/autograd/context/DistAutogradContext.cpp
Comment (pritamdamania87) [reply torch/csrc/distributed/autograd/context/DistAutogradContext.cpp]: We can modify this in follow up PRs to a map, right now we don't have an ID to put in this map.

**Comment by pritamdamania87** [comment]
@aazzolini As we discussed offline, when we pickle python objects, we can hook into the pickler and whenever we see tensors, we can attach them to the `send` autograd function appropriately.

File: torch/csrc/distributed/autograd/functions/sendrpc_backwards.h
Comment (ezyang) [inline torch/csrc/distributed/autograd/functions/sendrpc_backwards.h]: This gunna need a docblock

**Comment by ezyang** [comment]
> The purpose of this change is to validate the basic structure proposed here makes sense.

How do we actually validate this?

File: torch/csrc/distributed/rpc/python_functions.cpp
Comment (ezyang) [inline torch/csrc/distributed/rpc/python_functions.cpp]: This is not part of this PR, but I noticed it while reviewing the code here. Above, in line 38, there is a `opName.rfind("aten", 0) == 0`. That's a code smell to me: I generally don't think string matching in this way is a good idea, because it's easy to get confounded by cases like `"atenblah"`. Is there a reason we can't use `Symbol::fromQualString` and then test if the namespace is ATen? That seems semantically better to me.

File: torch/csrc/distributed/rpc/python_functions.cpp
Comment (ezyang) [inline torch/csrc/distributed/rpc/python_functions.cpp]: Also not part of this PR, but can we bound the try-catch block around the only scope which you actually expect to throw--the comment above implies `createStackForSchema` is the only thing that is currently throwing? This will help us from accidentally catching errors we don't intend to catch. It's very simple to do, while we're waiting for someone to actually refactor the downstream function in question. (Or you could just do that refactor?)

File: torch/csrc/distributed/rpc/python_functions.cpp
Comment (ezyang) [inline torch/csrc/distributed/rpc/python_functions.cpp]: Hmm. So if there is no current valid context, we just silently don't attach any autograd edges. That doesn't sound right to me: if there is no valid context, we had better *error* if we try to backpropagate.

Basically, if I don't do a context, and then I attend to rpc and backward through it, what error message do I get? That's something you should be testing for.

File: torch/csrc/distributed/autograd/utils.h
Comment (ezyang) [inline torch/csrc/distributed/autograd/utils.h]: This is actually a serious limitation, because it means that operations which take tensor lists like `torch.stack` silently don't work with the current implementation right now. You don't have to make it work immediately, but we should really give a good error message if you try to use something bad.

File: torch/csrc/distributed/autograd/context/dist_autograd_container.cpp
Comment (ezyang) [inline torch/csrc/distributed/autograd/context/dist_autograd_container.cpp]: nit: Not part of this PR, but we could have used a `TORCH_INTERNAL_ASSERT` here

Comment (ezyang) [reply torch/csrc/distributed/autograd/context/dist_autograd_container.cpp]: (You don't have to do anything to this code, I'm just letting you know in case you need to write something similar in the future)

File: torch/csrc/distributed/autograd/context/dist_autograd_container.cpp
Comment (ezyang) [inline torch/csrc/distributed/autograd/context/dist_autograd_container.cpp]: So, if I think about what the classic mistake that a user will make, it is that they will write distributed RPC code without actually setting the context. Is this the error message that will occur if this happens? If so, we could give more clear guidance about how to fix one's code here, rather than just stating what the internal violated invariant is.

File: torch/csrc/distributed/autograd/context/dist_autograd_container.cpp
Comment (ezyang) [inline torch/csrc/distributed/autograd/context/dist_autograd_container.cpp]: Nit: `find()` returns an iterator, which you can use to retrieve the element in question. The way you've written it here, you have to do *two* map lookups in the successful case when only one is necessary.

File: torch/csrc/distributed/autograd/context/dist_autograd_container.cpp
Comment (ezyang) [inline torch/csrc/distributed/autograd/context/dist_autograd_container.cpp]: Not part of this PR per se, but do we have a C++-side RAII guard for allocating and releasing thread local context? It seems like a logical thing that should be part of the C++ API.

File: torch/csrc/distributed/autograd/init.cpp
Comment (ezyang) [inline torch/csrc/distributed/autograd/init.cpp]: I'm... pretty sure this is leaking Python objects. Inspection of `functionToPyObject` reveals that it incref objects before returning. `reinterpret_borrow` is the "copying" constructor: it doesn't take ownership of the input.

Comment (ezyang) [reply torch/csrc/distributed/autograd/init.cpp]: (If I'm right, you should use `reinterpret_steal` here instead.)

**Review by ezyang — CHANGES_REQUESTED** [review]
Leaking Python objects, I think.

**Comment by ezyang** [comment]
To summarize the overall review: this looks fine, but it isn't really enough to validate if the design works or not (you implemented X, you wrote a unit test that shows if X works, but there isn't enough of the actual guts to built out to know if X is what you actually want). That's probably fine: this patch is pretty localized, so we should let it in ASAP and keep working on the real functionality. I am blocking land because I think it is leaking Python objects as it stands.

**Comment by pritamdamania87** [comment]
> How do we actually validate this?

The PR was more to validate I was building the custom autograd function appropriately and attaching the tensors.

File: torch/csrc/distributed/rpc/rpc_agent.cpp
Comment (zhaojuanmao) [inline torch/csrc/distributed/rpc/rpc_agent.cpp]: we may not be ready to attach C++ send/recv function for Pyper RPC integration yet.  But init_model_parallel() always init autograd and thus haveValidContext() is true. do we want to pass a disable_dist_auto_grad flag to  init_model_parallel() for now so that Pyper RPC integration can still use Python dist autograd version?

File: torch/csrc/distributed/autograd/context/dist_autograd_container.cpp
Comment (mrshenli) [reply torch/csrc/distributed/autograd/context/dist_autograd_container.cpp]: C++ API is non-goal for now, but let's add a TODO here in case we forgot.

File: torch/csrc/distributed/autograd/context/dist_autograd_container.cpp
Comment (mrshenli) [reply torch/csrc/distributed/autograd/context/dist_autograd_container.cpp]: Will this work for reentrant backward without nested context? (of course, reentrant backward support does not need to be part of the first release)

**Comment by mrshenli** [comment]
A code structure question: should `dist_autograd` be under the `rpc` folder? (if so, don't need to be addressed in this PR). I am asking because I was thinking about where we should put the `worker_id_t` type. This type should be shared by `rpc` and `dist_autograd`, and no other modules. But it does not make much sense to put this type directly under `distributed` folder. 

Do we foresee any use case for `dist_autograd` without using `rpc`? If not in near future, shall we merge `dist_autograd` into `rpc`?

File: torch/csrc/distributed/rpc/python_functions.cpp
Comment (mrshenli) [inline torch/csrc/distributed/rpc/python_functions.cpp]: This will print a lot of logs, I remember seeing ~20 overloaded versions of `aten::add`

File: torch/csrc/distributed/rpc/rpc_agent.cpp
Comment (mrshenli) [inline torch/csrc/distributed/rpc/rpc_agent.cpp]: I am not against adding it here in this PR. But in the future, we might want to decouple `RpcAgent` interface with autograd code. @satgera mentioned that eventually, we might want to have a separate rpc lib so that `RpcAgent` implementations can have a minimum set of dependencies.

**Review by mrshenli — APPROVED** [review]

**Comment by mrshenli** [comment]
Stamped to unblock, but please get another review from @ezyang and address @zhaojuanmao's comments before landing.

File: torch/csrc/distributed/rpc/python_functions.cpp
Comment (mrshenli) [reply torch/csrc/distributed/rpc/python_functions.cpp]: How many logs do you see when run tests locally?

**Review by ezyang — APPROVED** [review]
It's too early to tell if this all works, but the diff is not too long so let's move it along.

File: torch/csrc/distributed/autograd/context/dist_autograd_container.cpp
Comment (pritamdamania87) [reply torch/csrc/distributed/autograd/context/dist_autograd_container.cpp]: The user will probably invoke a specialized `torch.distributed.backward()` API. This will under the covers only have one autograd context. The engine itself can run multiple re-entrant backward passes under the hood, but we will associate all those backward passes with a single context.

File: torch/csrc/distributed/rpc/python_functions.cpp
Comment (pritamdamania87) [reply torch/csrc/distributed/rpc/python_functions.cpp]: Forgot to push an update, this was something I was testing.

File: torch/csrc/distributed/autograd/context/dist_autograd_container.cpp
Comment (pritamdamania87) [reply torch/csrc/distributed/autograd/context/dist_autograd_container.cpp]: Because we need to validate the worker_id below. If this is int16_t, we can pass a very large value (outside int16_t) and it will just wrap around.

File: torch/csrc/distributed/rpc/rpc_agent.cpp
Comment (pritamdamania87) [reply torch/csrc/distributed/rpc/rpc_agent.cpp]: `init_model_parallel` doesn't make `hasValidContext()` true. The user needs to explicitly use `with torch.distributed.autograd.context()` around a block of code to enable any of the functionality here.

File: test/test_dist_autograd.py
Comment (pritamdamania87) [reply test/test_dist_autograd.py]: I'm not sure.

File: torch/csrc/distributed/rpc/python_functions.cpp
Comment (pritamdamania87) [reply torch/csrc/distributed/rpc/python_functions.cpp]: I didn't push the latest version here, I've changed this to `VLOG(1)` so it will print only when verbosity is on.

File: torch/csrc/distributed/rpc/rpc_agent.cpp
Comment (pritamdamania87) [reply torch/csrc/distributed/rpc/rpc_agent.cpp]: Yes, I have re-worked this a bit in my follow up PR for the full forward pass.

**Comment by pritamdamania87** [comment]
> A code structure question: should dist_autograd be under the rpc folder? (if so, don't need to be addressed in this PR). I am asking because I was thinking about where we should put the worker_id_t type. This type should be shared by rpc and dist_autograd, and no other modules. But it does not make much sense to put this type directly under distributed folder.

Do we foresee any use case for dist_autograd without using rpc? If not in near future, shall we merge dist_autograd into rpc?

I think they should be in separate folders. For common stuff, we can put it under `distributed/common` and have both depend on them.

**Comment by pritamdamania87** [comment]
@ezyang @mrshenli https://github.com/pytorch/pytorch/pull/25527, which is built on top of this will hopefully give you a much more comprehensive picture.

**Pull request closed without merging**
